imap/jmap_api.c

```diff
@@ -2231,10 +2231,20 @@ HIDDEN void jmap_filter_parse(jmap_req_t *req, struct jmap_parser *parser,
             jmap_parser_invalid(parser, "operator");
         }
         arg = json_object_get(filter, "conditions");
-        json_array_foreach(arg, i, val) {
-            jmap_parser_push_index(parser, "conditions", i, NULL);
-            jmap_filter_parse(req, parser, val, unsupported, parse_condition, cond_rock, err);
-            jmap_parser_pop(parser);
+        if (json_is_array(arg)) {
+            json_array_foreach(arg, i, val) {
+                jmap_parser_push_index(parser, "conditions", i, NULL);
+                jmap_filter_parse(req, parser, val, unsupported,
+                                  parse_condition, cond_rock, err);
+                jmap_parser_pop(parser);
+            }
+        }
+        else if (arg) {
+            jmap_parser_invalid(parser, "conditions");
+        }
+        else {
+            /* XXX no conditions... is this an error or not? */
```

Comment:
Looking at https://jmap.io/spec-core.html#query I think this should report missing `conditions` as invalid. The type-specific query evaluators later are more lenient, e.g. the JMAP Email/query code certainly is. But it's good to filter out non-spec compliant queries.